tests/rptest/tests/data_transforms_test.py

```diff
@@ -0,0 +1,127 @@
+#
+# Use of this software is governed by the Business Source License
+# included in the file licenses/BSL.md
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0
+
+from rptest.clients.rpk import RpkTool
+from rptest.services.cluster import cluster
+from ducktape.utils.util import wait_until
+
+from rptest.tests.redpanda_test import RedpandaTest
+from rptest.clients.types import TopicSpec
+
+
+class WasmException(Exception):
+    def __init__(self, message):
+        self.message = message
+
+    def __str__(self):
+        return repr(self.message)
+
+
+class DataTransformsTest(RedpandaTest):
+    """
+    Tests related to WebAssembly powered data transforms
+    """
+    topics = [TopicSpec(partition_count=9), TopicSpec(partition_count=9)]
+
+    def __init__(self, test_context):
+        super(DataTransformsTest, self).__init__(test_context=test_context)
+        self._ctx = test_context
+        self._rpk = RpkTool(self.redpanda)
+
+    def _enable_wasm(self):
+        """
+        Enable data transforms and restart the cluster.
+        """
+        self._rpk.cluster_config_set("data_transforms_enabled", True)
+        self.redpanda.rolling_restart_nodes(self.redpanda.nodes)
+
+    def _deploy_wasm(self, name: str):
+        """
+        Deploy a wasm transform and wait for all processors to be running.
+        """
+        [input_topic, output_topic] = self.topics
+
+        def do_deploy():
+            self._rpk.deploy_wasm(name, input_topic.name, output_topic.name)
+            return True
+
+        wait_until(
+            do_deploy,
+            timeout_sec=30,
+            backoff_sec=5,
+            err_msg=f"unable to deploy wasm transform {name}",
+            retry_on_exc=True,
+        )
+
+        def is_all_running():
+            transforms = self._rpk.list_wasm()
+            transform = next((x for x in transforms if x.name == name), None)
+            if not transform:
+                raise WasmException(f"missing transform {name} from report")
+            for partition_id in range(0, input_topic.partition_count):
+                processor = next(
+                    (p
+                     for p in transform.status if p.partition == partition_id),
+                    None)
+                if not processor:
+                    raise WasmException(
+                        f"missing processor {name}/{partition_id} from report")
+                if processor.status != "running":
+                    raise WasmException(
+                        f"processor {name}/{partition_id} is not running: {processor.status}"
+                    )
+            return True
+
+        wait_until(
+            is_all_running,
+            timeout_sec=30,
+            backoff_sec=1,
+            err_msg=f"wasm transform {name} did not reach steady state",
+            retry_on_exc=True,
```

Comment:
question: Instead of raising an exception, why not return `False` and try again?

question: Is the purpose of the `wait_until` just to timeout in case the `list_wasm()` call hangs?

Reply:
I'm using `retry_on_exc=True` which retries on exceptions. The exceptions give better error messages than just returning false (I know why it's false). I need to retry on throws because the rpk commands throw if they fail

Reply:
Ok great!  And I think @NyaliaLui was making the same comment (which didn't show up when I was stepping through the commits, sorry for the noise).